import os 
import re
import glob

""" Snakerules for running 10x cellranger """

def _get_sample_ids(wildcards):
    """ extract out sample prefixes from fastqs to pass to cellranger
    i.e. given the following fastqs
    sample_name_1_S1_L001_R1_001.fastq.gz"
    sample_name_2_S2_L001_R1_001.fastq.gz"
    sample_name_3_S3_L001_R1_001.fastq.gz"

    return:
    sample_name_1,sample_name_2,sample_name_3
    """

    no_lane = "_S[0-9]+_[RI][12]_001\.fastq\.gz$"
    lane = "_S[0-9]+_L[0-9]{3}_[RI][12]_001\.fastq\.gz$"

    fq_dir = os.path.join(wildcards.data, "raw_data")          
    fqs = glob.glob(os.path.join(fq_dir, wildcards.sample + "*"))
    R1_fqs = [x for x in fqs if "R1" in x]
    R1_fqs = [os.path.basename(x) for x in R1_fqs]

    sample_ids = set()
    for fq in R1_fqs:
      if re.search(lane, fq) is not None:
        sample_id = re.sub(lane, "", fq)
        sample_ids.add(sample_id)

      elif re.search(no_lane, fq) is not None:
        # Novaseq style S1 flowcell
        sample_id = re.sub(no_lane, "", fq)
        sample_ids.add(sample_id)
      else:
        sys.exit("unable to parse out sample id from {}".format(fq))

    return ",".join(sample_ids)

def _get_sample_fq_paths(wildcards):
    """ generate list of fastqs that contain sample prefix 
    """
    
    sample_ids = _get_sample_ids(wildcards)
    sample_ids = sample_ids.split(",")

    fq_dir = os.path.join(wildcards.data, "raw_data")          
    input_fqs = set()
    for sample in sample_ids:
      
      r1_pattern = sample + "*" + "R1_001.fastq.gz"
      
      r1s = glob.glob(os.path.join(fq_dir, r1_pattern))
      r2s = [re.sub("_R1_001.fastq.gz$", "_R2_001.fastq.gz", x) for x in r1s] 
      reads = r1s + r2s
      for read in reads:
        input_fqs.add(read)
    
    print("running cellranger on the following fqs\nfastq\tsample", file = sys.stderr) 
    for fq in input_fqs:
      print("{}\t{}".format(fq, wildcards.sample), file = sys.stderr)

    return input_fqs

rule cellranger_count:
    """ note that cellranger count will fail if the output directory
    specified with ``--id`` exists prior to running. By default Snakemake 
    will generate directories listed in the input and output params, so to
    avoid this error this rule creates a dummy file "{sample}_complete.txt """

    input:
      _get_sample_fq_paths
    output:
      "{data}/10x_mrna/logs/{sample}_complete.txt"
    params:
      fqs = _get_sample_ids,
      outdir = "{data}/10x_mrna/",
      indir = MRNA_DATA,
      job_name = "{sample}.count",
      memory = "select[mem>4] rusage[mem=4]",
    log: "logs/10x_count/{sample}_count.out"
    threads: 2
    shell:
      """
      set -x
      echo {params.fqs}
      cellranger count \
          --id={wildcards.sample} \
          --fastqs={params.indir} \
          --sample={params.fqs} \
          --jobmode={LSF_TEMPLATE} \
          --project="haircut" \
          --maxjobs={MAX_10X_JOBS} \
          --transcriptome={MRNA_REF}

      mkdir -p {params.outdir}
      mv -u {wildcards.sample} {params.outdir}
      
      # test if output bam file produced by count
      bam="{params.outdir}{wildcards.sample}/outs/possorted_genome_bam.bam" 

      if [ -s $bam ] 
        then 
        echo "finished" > {output}
      fi
      """ 

